### PR TITLE
example/memcached-operator: implement services resource

### PR DESCRIPTION
**Pull request by kopiczko**
Tested with:

    $ k run --image=alpine --rm -it test /bin/sh

    If you don't see a command prompt, try pressing enter.
    / # printf "set mykey 0 60 4\r\ndata\r\n" | nc memcached0000 11211
    STORED
    / # nc memcached0000 11211
    get mykey
    VALUE mykey 0 4
    data
    END

### CODE REVIEW

File: example/memcached-operator/controller/resource/services.go [hunk 1]
```diff
@@ -2,39 +2,179 @@ package resource
...
-	s := &Services{}
-	return s, nil
+	d := &Services{
+		k8sClient: config.K8sClient,
+	}
+
+	return d, nil
 }
-func (s *Services) Name() string {
+func (d *Services) Name() string {
```
Comment (xh3b4sd) [inline example/memcached-operator/controller/resource/services.go]: Why `d` though? The resource is `service` and it was `s` which was just right IMO.

File: example/memcached-operator/controller/resource/services.go [hunk 2]
```diff
@@ -2,39 +2,179 @@ package resource
...
+		LabelSelector: key.LabelSelectorManagedBy,
+	})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	currentReplicas := len(managedServices.Items)
+	desiredReplicas := memcachedConfig.Spec.Replicas
+
+	logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("currentReplicas = %d desiredReplicas = %d", currentReplicas, desiredReplicas))
```
Comment (xh3b4sd) [inline example/memcached-operator/controller/resource/services.go]: This kind of logging is really not nice IMO. We have structured logging and here we then try to put structure into a string. Like why?

File: example/memcached-operator/controller/resource/services.go (hunk 1 above)
Comment (kopiczko) [reply example/memcached-operator/controller/resource/services.go]: Yes, thanks I just did bunch of substitutions and missed this detail.

**Comment by rossf7** [comment]
I hit a problem when testing with scaling the services resource. The clusterIP in the Service spec is immutable but we try to set it to an empty string. If we need to update I think we'll need to do a Patch without the clusterIP.

```
{"caller":"github.com/giantswarm/operatorkit/controller/controller.go:311","event":"update","level":"error","message":"stop reconciliation due to error","object":"/apis/example.giantswarm.io/v1alpha1/namespaces/default/memcachedconfigs/test-memcached","stack":"[{/Users/ross/.gvm/pkgsets/go1.10/giantswarm/src/github.com/giantswarm/operatorkit/controller/controller.go:434: } {/Users/ross/.gvm/pkgsets/go1.10/giantswarm/src/github.com/giantswarm/operatorkit/example/memcached-operator/controller/resource/services.go:60: } {/Users/ross/.gvm/pkgsets/go1.10/giantswarm/src/github.com/giantswarm/operatorkit/example/memcached-operator/controller/resource/services.go:116: } {Service \"memcached0000\" is invalid: [metadata.resourceVersion: Invalid value: \"\": must be specified for an update, spec.clusterIP: Invalid value: \"\": field is immutable]}]","time":"2018-05-15 11:35:59.043"}
```

**Comment by kopiczko** [comment]
@rossf7 I noticed. I just addressed that in https://github.com/giantswarm/operatorkit/pull/211/commits/40344fce7f2c76e13ee8b5ea7007ae88349eb129

File: example/memcached-operator/controller/resource/services.go [hunk 3]
```diff
@@ -2,39 +2,189 @@ package resource
...
+		_, err = s.k8sClient.CoreV1().Services(desired.Namespace).Create(desired)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("service %s/%s created", desired.Namespace, desired.Name))
+	} else {
+		// Service.spec.clusterIP is immutable and filled by the
+		// Kubernetes when not provided.
+		desired.Spec.ClusterIP = current.Spec.ClusterIP
```
Comment (kopiczko) [inline example/memcached-operator/controller/resource/services.go]: @rossf7 this is the line. I also update deployments resource to be aligned.

**Comment by kopiczko** [comment]
Closing. With introduction of `resource` package this will be soon outdated anyway.

**Pull request closed without merging**